Add LCPO method

epretti opened the pull request:
Adds LCPO (linear combination of pairwise overlaps) as discussed in #4746.

- [X] Reference platform
- [X] CPU platform
- [x] Common platform
- [X] Parameters for Amber files
- [x] Parameters for Topology + XML
- [x] Optimize common platform

peastman commented on docs-source/usersguide/application/02_running_sims.rst:
```diff
@@ -983,6 +983,16 @@ Debye-Huckel screening parameter\ :cite:`Srinivasan1999`:
+approximation :cite:`Weiser1999`:
+::
+
+    system = prmtop.createSystem(implicitSolvent=OBC2, gbsaModel='LCPO')
```
How about calling the argument `sasaMethod`?  The name `gbsaModel` suggests something much more general.  This option doesn't effect the GB part of the calculation, and also doesn't affect how energies depend on the SASA.  It's just the method used to  estimate the SASA.

epretti replied:
I can change it if you think it is OK to do so; I agree that `sasaMethod` makes more sense but I didn't want to break the existing API so I kept the existing name.

peastman replied:
> Another option is to rename the AmberPrmtopFile argument to sasaMethod and make gbsaModel a deprecated alias for it.

That sounds like a good choice.

peastman commented on docs-source/usersguide/theory/02_standard_forces.rst:
```diff
@@ -462,6 +462,42 @@ where :math:`r_i` is the atomic radius of particle *i*\ , :math:`r_i` is
+
+Note that the coefficients used in LCPOForce should be set to the product of the
+coefficients in the LCPO manuscript :cite:`Weiser1999` with the desired surface
+energy.
```
This sentence is a little confusing.  I'm not quite sure what it means.

peastman commented on openmmapi/include/openmm/LCPOForce.h:
```diff
@@ -0,0 +1,144 @@
+     * @param p1      the product of the surface energy with the value of the LCPO parameter P1 for the particle
+     * @param p2      the product of the surface energy with the value of the LCPO parameter P2 for the particle
+     * @param p3      the product of the surface energy with the value of the LCPO parameter P3 for the particle
+     * @param p4      the product of the surface energy with the value of the LCPO parameter P4 for the particle
```
Would it make sense to pass the surface area scale as a separate parameter rather than asking the user to premultiply by it?  That better matches how it's described in the paper and specified in Amber.

epretti replied:
I can change this.  Should it be a per-particle parameter?  For Amber, the surface energy is a single parameter for the whole system, although there is an undocumented and possibly incomplete feature in the source code that lets you define different energies per amino acid.

peastman replied:
If it's usually specified as a single number, that's how I'd do it.  There's no loss of generality, since you can always absorb variation between particles into the other parameters.  It's just to make the interface match what people expect.

We also could make it a context parameter, allowing you to inexpensively change it.  I don't know whether there are cases where someone would want to do that.

epretti commented:
> Did I understand correctly, and each triplet gets processed six times?

That's correct; all of the permutations are processed for each triplet.  The calculation is asymmetric so that was the most natural way for me to write it.

> If so, is it necessary to do that (because the calculation is different each time)?  Or could you process each one only once and multiply the result by 6?

I'd have to do separate calculations for each permutation in the inner loops, but it's possible that some time can be saved building the neighbor list like this, and perhaps there would be an opportunity to reuse a few values between calculations for the different permutation.  I've already found that recomputing the distance between i and k to check if they are neighbors is faster than scanning the neighbor list.  I will test this out as well, because I think it would then be possible to do half as many of those distance checks in the inner loop.

epretti commented:
@peastman You can take a look at the common platform implementation now if you want.

I have been spending some time trying to optimize this.  I have a few test systems of various sizes, and the current performance I am getting is as follows (CUDA single precision, ns/day, median of 5 runs 10s each) compared to ACE and the slow CustomManyParticleForce implementation of LCPO for reference:

| System | Atoms | ACE | LCPO (custom) | *LCPO (this)* |
| --- | --- | --- | --- | --- |
| Chignolin | 138 | 1823 | 882 | 1235 |
| DHFR | 2489 | 1170 | 225 | 759 |
| Nucleosome | 25095 | 157 | 30 | 118 |

Right now it seems to be about 25-35% slower than just using ACE depending on system size (for the nonbonded options selected), but a lot faster than the CustomManyParticleForce implementation.  Let me know if you see any obvious opportunities for further optimization.  In case you want to profile it yourself, the benchmark script and inputs I have been using are here:

[benchmark.tar.gz](https://github.com/user-attachments/files/23781835/benchmark.tar.gz)

peastman commented:
Did you try ordering `condensedPos` in the other way (the x coordinates of all atoms, followed by the y coordinates of all atoms, etc.)?  It might lead to better coalescing of memory access in `findNeighbors` and `computeInteraction`.

peastman commented:
Here is another possible optimization.  In the inner loop of `computeInteraction`, you first check the distance between i and k, which is fairly quick to do.  If it's beyond the cutoff, you skip the rest of the calculation, which is the much more expensive part.  On DHFR, less than 50% of k atoms pass the check, so you can usually skip the expensive calculation.

Except of course on a GPU, you can't.  If even one thread passes the check, every thread in the warp pays the cost of the expensive calculation.  Thread divergence leads to a lot of wasted calculation.

You might be able to improve the performance by restructuring the loop like this.

1. Split the loop into smaller blocks, maybe around 10 atoms.
2. For each block, first do the cutoff check on every atom.  For any atom that passes the check, store `k`, `kParams`, and `ikDelta` into a buffer.
3. Loop over the atoms you stored in the buffer and do the expensive calculation on them.

This will hopefully produce less thread divergence and better efficiency.

You can use shared memory for the buffer.  It's very fast to access and you don't use it for anything else.  The larger the blocks are, the less variation you'll get between threads in how many atoms are stored into the buffer.  But if you make them too large, you'll run out of shared memory and force it to reduce occupancy.  You can figure out a reasonable block size based on those considerations.

peastman commented:
You also could make `condensedPos` a `real4`.  Put the radius into the fourth element.  That way you only need one load for the whole thing, which really should be more efficient when you aren't getting good coalescing.  In `findNeighbors` you won't need to access `parameters` at all, since the only thing you use it for is the radius.  And in `computeInteraction` you'll only need to load it for the k atoms that pass the distance check.

You also could split up `parameters` into a pair of real2 arrays, one with `radius` and `p2`, and the other with `p3` and `p4`.  The reason is that in the inner loop of `computeInteraction`, you never use `p2`, and `radius` would already have been loaded from `condensedPos`.  So you wouldn't have to load as much data for each k atom.

epretti commented:
Here's where I'm at with this feature:

*Parameter assignment:*

I've added a heuristic approach for assigning the LCPO parameters from a Topology.  The two non-trivial cases are:

1. Determining whether an N with 3 bonds is sp3 (trigonal pyramidal) or sp2 (trigonal planar, either positively charged with a double bond and two single bonds, part of an aromatic ring, or most commonly, part of an amide).
2. Determining whether an O with 1 bond is part of an ordinary carbonyl group (and should have sp2 hybridization), or part of a carboxylate ion (such that it gets special LCPO parameters).

(1) is handled by checking the hybridization of C atoms bonded to an N atom in question.  I think this may fail for some cases like azides where N is bonded to itself.  However, the new test case I added covering all of the standard amino acids and nucleotides correctly assigns all of these N atoms.

(2) is handled by specifically identifying the topology of a carboxylate.  This correctly identifies all of the carboxylate O atoms, but it doesn't actually match Amber!  This is because Amber reuses the carboxylate `O2` atom type it checks for in this case for oxygens attached to phosphate groups, so it ends up assigning the carboxylate type to them.

I'm not sure what behavior we should implement here.  The LCPO paper doesn't mention doing this, such that I think assigning the carboxylate type to phosphate O atoms is wrong.  I looked at what NAMD does; NAMD reading Amber inputs emulates the Amber behavior, while when it reads CHARMM inputs, it will end up assigning an sp3 atom type to these O atoms (whether this is intended behavior is also unclear since it is the result of a default/fallthrough case).

*Trying to optimize further:*

> > Did you try ordering `condensedPos` in the other way (the x coordinates of all atoms, followed by the y coordinates of all atoms, etc.)? It might lead to better coalescing of memory access in `findNeighbors` and `computeInteraction`.
> 
> I had already tried that and (was a bit surprised to find) it didn't help. But I think I did a few other optimizations since then, so I will double-check to see if it has any effect on the current version.

Confirmed that this doesn't help the current version.

> You might be able to improve the performance by restructuring the loop like this.
> 
> 1. Split the loop into smaller blocks, maybe around 10 atoms.
> 
> 2. For each block, first do the cutoff check on every atom.  For any atom that passes the check, store `k`, `kParams`, and `ikDelta` into a buffer.
> 
> 3. Loop over the atoms you stored in the buffer and do the expensive calculation on them.

I already tried something like this before with the old neighbor checking approach I was using and it didn't help.  I've tested it again with the current cutoff-based approach.  I've now tried it with different neighbor block sizes, thread block sizes, and shared memory access patterns.  I couldn't find any combination of parameters that wasn't slower than not doing this.

> You also could make `condensedPos` a `real4`. Put the radius into the fourth element.

Inexplicably this slows down `findNeighbors` by 50% for me; I can't understand why!  This offsets any benefit to `computeInteraction` in the test cases I have.

epretti commented:
Here's a patch.  The version I had stashed also has the suggested splitting of `parameters` into separate arrays, which I recall didn't appear to do much either way.

[condensedPosReal4.patch](https://github.com/user-attachments/files/23940546/condensedPosReal4.patch)

peastman commented:
I don't see the slowdown you describe, though I also don't see any significant benefit.  According to the compute profiler, the median duration of `findNeighbors` decreases from 39.9 to 39.0.  The median duration of `computeInteraction` decreases from 59.9 to 59.3.  Those differences are small enough that I'm not convinced they aren't just noise.
